src/main/scala/async/Async.scala

```diff
@@ -150,108 +151,59 @@ object Async:
-      //   var found = false
-
-      //   val listener = new Listener[T]:
-      //     def tryLock() =
-      //       val result = k.tryLock()
-      //       // If tryLock is called, the source has - to the best of its
-      //       // knowledge - an item available. But the upstream listener k
-      //       // refuses to take any -> we assume there would have been one.
-      //       if result == Listener.Gone then found = true
-      //       result
-
-      //     def release(until: Listener.ReleaseBoundary) = k
-
-      //     def complete(data: T) =
-      //       k.complete(data)
-      //       found = true
-      //   end listener
-
-      //   while it.hasNext && !found do
-      //     it.next.poll(listener)
-      //   found
-
       def onComplete(k: Listener[T]): Unit =
-        val listener = new Listener.ForwardingListener[T](this, k) with Listener.LockingListener { self =>
-          var foundBefore = false
-
-          val listenerPartial = new Listener.NumberedSublock {
-            def number: Long = self.number
-            def tryLock() =
-              self.lock()
-              val result =
-                if foundBefore then
-                  self.unlock()
-                  Listener.Gone
-                else
-                  val result = k.tryLock()
-                  if result == Listener.Gone then self.unlock()
-                  result
-              end result
-              // as soon as we once return None, we are completed -> drop everywhere
-              if result == Listener.Gone then sources.foreach(_.dropListener(self))
-              result
-          } // end listenerPartial: NumberedSublock
-
-          def tryLock() =
-            if foundBefore then Listener.Gone
-            else listenerPartial
-
-          def release(until: Listener.ReleaseBoundary) =
-            if until != listenerPartial then
-              self.unlock()
-              k
-            else null
-
-          def complete(data: T) =
-            k.completeNow(data)
-            foundBefore = true
-            self.unlock()
-            sources.foreach(_.dropListener(self))
+        val listener = new Listener.ForwardingListener[T](this, k) with NumberedLock with Listener.ListenerLock with Listener.PartialLock { self =>
+          val lock = self
+
+          var found = false
+          inline def heldLock = if k.lock == null then Listener.Locked else this
+
+          /* == PartialLock implementation == */
+          // Note that this is bogus if k.lock is null, but we'll never use it if it is.
+          val nextNumber = if k.lock == null then -1 else k.lock.selfNumber
+          def lockNext() =
+            val res = k.lock.lockSelf(selfSrc)
+            if res == Listener.Gone then
+              found = true // This is always false before this, since PartialLock is only returned when found is false
+              sources.foreach(_.dropListener(this)) // same as dropListener(k), but avoids an allocation
```

Comment:
The problem is that this code only catches Gone in the immediately following downstream level. If only two levels later, the listener is Gone, we only get to know this in release. Outsourcing the lock iteration, the only way I can come up to solve this, is by moving the `if res == Listener.Gone then` body to a `if until != null then` in release.

Reply:
This is not correct, because the listener can be released even if downstream is not `Gone` (e.g. as the other branch of `lockBoth`).